Multiplayer Component Version Mismatch Detection

AMZN-Gene opened the pull request:
- Server will detect when a connecting client has different multiplayer components and will send back a list of all its components to compare what has been modified, added, or removed.
- A connection with a mismatch will be disconnected once all the components are synced up
- After disconnecting the client can be updated to match the server and reconnect without the server shutting down
- Unit Tests

![image](https://user-images.githubusercontent.com/32776221/199592999-8af1f83d-d127-46cd-8c98-fa792a1be4e3.png)

Part 1 implementation for #9669 and https://github.com/o3de/sig-network/issues/81 
Upcoming PR
- Events when version mismatch occurs so that game-specific logic can be ran
- On-screen debug information
 
## How was this PR tested?
1. Modify a multiplayer component, add a multiplayer component on client, add a multiplayer component on server and ensure error logs are seen containing information about which components are mismatched.

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/Components/MultiplayerComponentRegistry.cpp:
```diff
@@ -70,8 +75,31 @@ namespace Multiplayer
+        return m_componentVersionHashes;
+    }
+
     void MultiplayerComponentRegistry::Reset()
```
Should this clear `m_componentVersionHash `as well?

AMZN-Gene replied:
thank you!

AMZN-puvvadar commented on Gems/Multiplayer/Code/Include/Multiplayer/Components/MultiplayerComponentRegistry.h:
```diff
@@ -67,11 +67,26 @@ namespace Multiplayer
     private:
         NetComponentId m_nextNetComponentId = NetComponentId{ 0 };
         AZStd::unordered_map<NetComponentId, ComponentData> m_componentData;
+        AZ::HashValue64 m_componentVersionHash = AZ::HashValue64{ 0 };
```
`m_componentVersionHash` and `m_componentVersionHashes` are very similar names despite representing different things. I'd consider renaming the former to something more descriptive.

AMZN-Gene replied:
Renamed! Good call, the RFC referred to it as a "holistic" hash, and the Getter method refers to it that way as well... fixed!

AMZN-puvvadar commented:
The approach looks good generally. One question I have is do we want to send the full dataset of hashes for the mismatch? That can certainly be useful for debugging but I wonder if that's too descriptive in a production environment.

For example, should a client get that much information about server hashes or should it just get that it is behind the server version and needs to upda[te?

AMZN-Olex commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -842,6 +873,91 @@ namespace Multiplayer
+            else
+            {
+                // Connected machine is using a multiplayer component we don't have
+                AZLOG_ERROR(
```
I am wondering if this should be an ebus to report on? This way it could be easy to add it to the Editor for example, if there is some odd reason when this fails.

AMZN-Gene replied:
Yes! Events coming in the next PR along with on-screen debug information, and unit tests.

AMZN-Olex commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -842,6 +873,91 @@ namespace Multiplayer
+
+        // Disconnect and clear out the list of components in case the user fixes up their end and tries to reconnect
+        theirMultiplayerComponents.clear();
+        connection->Disconnect(DisconnectReason::VersionMismatch, TerminationEndpoint::Local);
```
Hmm, is there an ebus/event here that a user can sign up and detect this version mismatch? Perhaps, we should pass some extra data here as well for debug-ability?

AMZN-Olex replied:
For example, I am curious what would be needed to expose to the Editor? Looking into the logs can often difficult.

AMZN-Gene replied:
Yup! Events coming in the next PR
I wasn't planning anything special just for the editor, but if the mp viewport debug message system is enabled, I'll display a popup saying there's a mismatch. Also the event could be used for game-specific logic (like having the user download the latest version).

AMZN-Gene commented on Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerTypes.h:
```diff
@@ -113,6 +113,20 @@ namespace Multiplayer
         Enabled
     };
 
+        //! The maximum number of component version hashes we want to stuff into a single packet
```
fixed!

AMZN-Gene commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -600,6 +600,37 @@ namespace Multiplayer
             }
         }
+
+        // Make sure the client that's trying to connect has the same multiplayer components
```
Thought about that as well; i think the only time i'd disable it is if the system itself was broken and wasn't reliably finding mismatches. Excluding not working as expected, is there a reason you'd want to run a client and server with different components?

AMZN-Gene replied:
Fixed! Added sv_versionMismatch_autoDisconnect which when false will still allow players to join even if there is a mismatch.

AMZN-Gene commented:
> The approach looks good generally. One question I have is do we want to send the full dataset of hashes for the mismatch? That can certainly be useful for debugging but I wonder if that's too descriptive in a production environment.
> 
> For example, should a client get that much information about server hashes or should it just get that it is behind the server version and needs to update?

I was thinking if it occurred it would be super useful for debugging if players uploaded their logs to the developer. You're thinking the component names could be maliciously used for hacking? I could make a cvar to bypass sending individual hashes and just call disconnect(versionMismatch). Although there's also talks of a cvar to not disconnect mismatch, in which case, I could send an different less-information-mismatch packet.

AMZN-puvvadar commented:
> > The approach looks good generally. One question I have is do we want to send the full dataset of hashes for the mismatch? That can certainly be useful for debugging but I wonder if that's too descriptive in a production environment.
> > For example, should a client get that much information about server hashes or should it just get that it is behind the server version and needs to update?
> 
> I was thinking if it occurred it would be super useful for debugging if players uploaded their logs to the developer and they could see what component went wrong. You're thinking the component names could be maliciously used for hacking? I could make a cvar to bypass sending individual hashes and just call disconnect(versionMismatch). Although there's also talks of a cvar to not disconnect on mismatch, in which case, I could create a new less-information-mismatch packet.



> > The approach looks good generally. One question I have is do we want to send the full dataset of hashes for the mismatch? That can certainly be useful for debugging but I wonder if that's too descriptive in a production environment.
> > For example, should a client get that much information about server hashes or should it just get that it is behind the server version and needs to update?
> 
> I was thinking if it occurred it would be super useful for debugging if players uploaded their logs to the developer and they could see what component went wrong. You're thinking the component names could be maliciously used for hacking? I could make a cvar to bypass sending individual hashes and just call disconnect(versionMismatch). Although there's also talks of a cvar to not disconnect on mismatch, in which case, I could create a new less-information-mismatch packet.

That's my concern yeah. I think rather than relying on users sending logs it'd be better to send a packet back to the server and then let the developer decide what they want to do with that data.

AMZN-Gene commented:
> That's my concern yeah. I think rather than relying on users sending logs it'd be better to send a packet back to the server and then let the developer decide what they want to do with that data.

Fixed! Added sv_versionMismatch_sendAllComponentHashesToClient. Set to false and the client will only hear about the mismatch, but not get any details on the specific components.

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -842,6 +871,125 @@ namespace Multiplayer
+            }
+            else
+            {
+                AZLOG_ERROR(
```
NP: Not sure it's worth saying "please file a bug" here.

Also could this be construed as not trying to accept a specific mismatched connection with the cvar off? That use case seems rare though.

AMZN-Gene replied:
Maybe an assert is better here? Basically if that code ever gets hit... the code is wrong and a developer needs to fix it.

AMZN-Gene replied:
_"Also could this be construed"_
Ah I see, the logic flow was weird to read with the if/else-if/else. Updated so it's just an if/else and then another if/else inside the else, which is easier on the eyes. Lemme know!

AMZN-Gene replied:
Replaced with assert.

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -1485,6 +1638,32 @@ namespace Multiplayer
+
+            if (componentMismatchPacket.GetComponentVersions().full())
+            {
+                // we've hit the packet limit, send the current packet, and begin filling up another
```
This alone makes the "send no manifest" cvar worth it. On a bulky mismatch, it'll keep things quick.

One thing you can consider here is doing one large packet of dynamic size where you store the size of the buffer followed by the buffer. That way you'll leverage AzNetworking's large packet fragmentation. Not strictly necessary but something to think about.

AMZN-Gene replied:
sv_versionMismatch_sendAllComponentHashesToClient = false stops sending all information to the client. Think there should be another cvar that eliminates sending all the components to both client and server?

AMZN-Gene replied:
Ah, I wasn't aware I could send non-fixed size vectors. I didn't want to assume a large fixed size since there really isn't any cap on the number of components a game can have. Dynamic packet size would let me send just one packet, point me at any existing packet example if you get a chance.

AMZN-puvvadar replied:
I don't think it's been attempted but look at AzContainerSerializers.h. Non-fixed size containers have an implementation along with non-fixed size strings.

The question on sv_versionMismatch_sendAllComponentHashesToClient is an implementation detail for you to consider.

AMZN-Gene replied:
Love it! I'm using a dynamically sized map and the code is a lot cleaner. Check out AzContainerSerializers.h, I updated it to include pairs since that's needed for the map iterator, but otherwise it worked without a problem.

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Header.jinja:
```diff
@@ -533,6 +533,7 @@ namespace {{ Component.attrib['Namespace'] }}
         bool HasController() const override;
         Multiplayer::MultiplayerController* GetController() override;
         const Multiplayer::MultiplayerController* GetController() const override;
+        static AZ::HashValue64 GetVersionHash() { return s_versionHash; }
```
Really try to keep implementations out of headers.. even trivial guys like this, as it has a tendency to get out of control quickly in a large codebase like what we're dealing with..

AMZN-Gene replied:
Thank you, fixed

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/Components/MultiplayerComponentRegistry.h:
```diff
@@ -67,11 +67,26 @@ namespace Multiplayer
 
+        //! Returns the combined hashes of all the multiplayer components creating a single holistic version hash that can be compared between the server and client app.
+        //! @return a 64-bit hash value representing of all of the multiplayer components
+        AZ::HashValue64 GetMultiplayerComponentVersionHolisticHash() const;
```
Holistic seems a bit weird too 😆 CombinedHash or AggregateHash might read a bit better? Perhaps even ComponentRegistryHash, although that is definitely a bit more esoteric..

AMZN-puvvadar replied:
Maybe just GetMultiplayerSystemVersionHash? I think the focus here should be on describing what it is not how it's formed.

AMZN-Gene replied:
I like it! Name changed to SystemHash.

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/Components/MultiplayerComponentRegistry.h:
```diff
@@ -67,11 +67,26 @@ namespace Multiplayer
 
+        //! Returns the combined hashes of all the multiplayer components creating a single system version hash that can be quickly compared between the server and client app.
+        //! @return a 64-bit system-wide hash value representing of all of the multiplayer components
+        AZ::HashValue64 GetSystemVersionHash() const;
```
nit: GetMultiplayerSystemVersionHash or GetCombinedMultiplayerVersionHash? 

Feel that GetSystemVersionHash is a little too brief and not as self-describing as we could make it

AMZN-puvvadar replied:
I would say GetMultiplayerSystemVersionHash between the two. Personally I find the namespace and class qualifies GetSystemVersionHash reasonably well enough.

AMZN-Gene replied:
Thanks for the feedback. Totally, both of the same thoughts crossed my mind, but I went with the less redundant "GetSystemVersion". The namespace and class covers the base nicely. Plus the calls for an outsider look like...
_GetMultiplayerComponentRegistry()->GetSystemVersionHash()_

lmbr-pip commented on Gems/Multiplayer/Code/Source/AutoGen/Multiplayer.AutoPackets.xml:
```diff
@@ -11,17 +11,22 @@
         <Member Type="Multiplayer::LongNetworkString" Name="map" />
     </Packet>
 
+    <Packet Name="SyncComponentMismatch" HandshakePacket="true" Desc="Sends a vector of multiplayer components with their name and version hashes in order to compare which particular components are mismatched. Note: Marked as a handshake packet so it can be sent back to the client without having received the server acceptance packet.">
```
Consider changing the name and description so that its clear this is synching component hashes.

Unclear from desc if this is all the hashes or just the ones which are different.

AMZN-Gene replied:
Updated the description to mention that all component information is sent in order to compare. Both the server and client receive this packet from one another. Wondering about the name change though, since I'm more concerned about the event (mismatch) than the under-lying hashes. Maybe "component" is making it confusing; I'm thinking "VersionMismatch" instead of "SyncComponentMismatch" would be a better name. Like a "version mismatch" event just occurred.

lmbr-pip commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -679,6 +662,52 @@ namespace Multiplayer
+        }
+
+        // Make sure the client that's trying to connect has the same multiplayer components
+        if (GetMultiplayerComponentRegistry()->GetSystemVersionHash() != packet.GetMultiplayerComponentVersionHash())
```
Is this the right check? These should compare "system hash" (ours) vs "system hash" theirs right? Naming makes that unclear IMHO

AMZN-puvvadar replied:
Second @lmbr-pip 's sentiment.

AMZN-Gene replied:
good called! changed to packet.GetSystemVersionHash

lmbr-pip commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -855,6 +884,104 @@ namespace Multiplayer
+            {
+                // Connected machine is using a multiplayer component we don't have
+                AZLOG_ERROR(
+                    "Multiplayer component mismatch! We're missing a component with version hash 0x%llx. "
```
nit: avoid personal pronouns in error messages. Unclear to reader who "we" is.

AMZN-Gene replied:
Fixed!

lmbr-pip commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -855,6 +884,104 @@ namespace Multiplayer
+                if (!theyHaveComponent)
+                {
+                    AZLOG_ERROR(
+                        "Multiplayer component mismatch! We have a component named %s which the connected application is missing!",
```
nit: personal pronouns in error message

AMZN-Gene replied:
Fixed. Replaced "We" with "This application"

lmbr-pip commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -855,6 +884,104 @@ namespace Multiplayer
+                if (m_originalConnectPackets.contains(connection->GetConnectionId()))
+                {
+                    // DANGER: We're accepting the player connection even though there's a component mismatch
+                    AZLOG_WARN("Multiplayer component mismatch was found but we're allowing the player to connect anyways. Please set "
```
```suggestion
                    AZLOG_WARN("Multiplayer component mismatch was found. Server configured to allow the player to connect anyways, which may result in undefined behavior. Set "
```

AMZN-Gene replied:
<3

AMZN-puvvadar reviewed (approved):
With pending resolution of issues around naming/comments in mind, this looks good to me.

AMZN-Gene commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -97,7 +97,9 @@ namespace Multiplayer
     AZ_CVAR(bool, bg_multiplayerDebugDraw, false, nullptr, AZ::ConsoleFunctorFlags::Null, "Enables debug draw for the multiplayer gem");
     AZ_CVAR(bool, cl_connect_onstartup, false, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Whether to call connect as soon as the Multiplayer SystemComponent is activated.");
-    
+    AZ_CVAR(bool, sv_versionMismatch_autoDisconnect, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Should the server automatically disconnect a client that is attempting connect who is running a build containing different/modified multiplayer components.");
```
I'm conflicted on this; would users care how the mismatch is detected (using a hash), or simply that's there was a mismatch?

AMZN-Gene replied:
Seeing the related review comment below about the packet name "SyncComponentMismatch". What if I changed the packet name to "VersionMismatch"? Trying to lower the user's burden of having to know how the mismatch is detected (via hashes) vs just knowing that there is a mismatch.

AMZN-Gene commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -842,6 +871,127 @@ namespace Multiplayer
+            {
+                // Connected machine is using a multiplayer component we don't have
+                AZLOG_ERROR(
+                    "Multiplayer component mismatch! We're missing a component with version hash 0x%" PRIx64 ". "
```
good call. fixed to "this application"

AMZN-Gene commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -97,7 +97,9 @@ namespace Multiplayer
     AZ_CVAR(bool, cl_connect_onstartup, false, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Whether to call connect as soon as the Multiplayer SystemComponent is activated.");
-    
+    AZ_CVAR(bool, sv_versionMismatch_autoDisconnect, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Should the server automatically disconnect a client that is attempting connect who is running a build containing different/modified multiplayer components.");
+    AZ_CVAR(bool, sv_versionMismatch_sendAllComponentHashesToClient, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Should the server send all its individual multiplayer component version hashes to the client when there's a mismatch? Upon receiving the information, the client will print the mismatch information to the game log. This is good for debugging during development, but you may w ...
```
thanks, fixed!

AMZN-Olex reviewed (approved):
Looks good!
